### PR TITLE
Color Customization V1

### PR DESCRIPTION
**Pull request by lanzhenw**
Added color scheme set up feature:

- User can set up the color setting in the UI to specify customized rules to display embedded document object's color setting. 
- User can apply color scheme rules in the SDK by defining `session.color_scheme`
- User can save color scheme of the dataset to `dataset.app_config`

In loading, session color scheme >> dataset.app_config >> default app setting

Setting colors in the UI:
https://user-images.githubusercontent.com/17770824/234972360-44ea9c70-6a5e-4552-b7f3-5f231329c8f5.mp4

Setting the colors in the SDK:
https://user-images.githubusercontent.com/17770824/234973226-6a65fbc2-46db-428d-91f6-995b543c0069.mp4

Save to dataset:
https://user-images.githubusercontent.com/17770824/234973275-374a1c14-4977-4dbf-8814-f098f6121cca.mp4

## How to test it?
```
import fiftyone as fo
import fiftyone.zoo as foz


dataset = foz.load_zoo_dataset("quickstart")
session = fo.launch_app(dataset)
    
color1 = fo.ColorScheme(color_pool=['blue', 'green', 'yellow', 'red'], customized_color_settings=[{'field': 'ground_truth', 'useFieldColor': True, 'fieldColor': 'pink', 'useLabelColor': True, 'labelColors': [{'name': 'dog', 'color': 'purple'}, { 'name': 'cat', 'color': 'pink'}]}])
session.color_scheme = color1
```

## How is this patch tested? If it is not, please explain why.

(Details)

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Review by manivoxel51 — COMMENTED** [review]
nice work Lanny! leaving some thoughts - still reviewing / testing

File: app/packages/core/src/components/Sidebar/Sidebar.tsx [hunk 1]
```diff
@@ -436,6 +437,13 @@ const InteractiveSidebar = ({
     () => new Controller({ minHeight: 0 })
   );
 
+  const customizeColor = useRecoilValue(fos.customizeColorSettings);
+  const colorFields = useRecoilValue(fos.customizeColorFields);
+  const tempColor = useRecoilValue(tempColorSetting);
+  console.info(customizeColor);
+  console.info(colorFields);
+  console.info("tempColor", tempColor);
+
```
Comment (benjaminpkane) [inline app/packages/core/src/components/Sidebar/Sidebar.tsx:446]: Unused code?

**Review by benjaminpkane — COMMENTED** [review]
Looks really good! Still playing with functionality. We can sync up today to talk backend details

File: app/yarn.lock [hunk 2]
```diff
@@ -4376,7 +4376,9 @@ __metadata:
     prettier: 2.2.1
     re-resizable: ^6.8.0
     react: 18.1.0
+    react-color: ^2.19.3
     react-dom: 18.1.0
+    react-draggable: ^4.4.5
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/yarn.lock:4381]: Yes, for making the modal draggable.

File: app/packages/core/src/components/ColorModal/FieldSetting.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  const sessionColorState = useRecoilValue(fos.sessionColorScheme);
+  const fullSetting = sessionColorState.customizedColorSettings;
+  const pool = sessionColorState.colorPool;
+  const coloring = useRecoilValue(fos.coloring(false));
+  const color = getColor(pool, coloring.seed, path);
+
+  const defaultColor =
+    coloring.pool[Math.floor(Math.random() * coloring.pool.length)];
+  const expandedPath = useRecoilValue(fos.expandPath(path!));
+  const VALID_COLOR_ATTRIBUTE_TYPES = [BOOLEAN_FIELD, INT_FIELD, STRING_FIELD];
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/FieldSetting.tsx:41]: nit: move this to a constant outside

File: app/packages/core/src/components/ColorModal/FieldSetting.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  const hideFieldColorPicker = (e) => {
+    if (
+      e.target.id != "twitter-color-container" &&
+      !e.target.id.includes("input")
+    ) {
+      setShowFieldPicker(false);
+    }
+  };
+
+  // on initial load, if the tem
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/FieldSetting.tsx:76]: `tem`

File: app/packages/core/src/components/ColorModal/JSONViewer.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    fos.appConfigOption({ key: "multicolorKeypoints", modal: false })
+  );
+  const showSkeleton = useRecoilValue(
+    fos.appConfigOption({ key: "showSkeletons", modal: false })
+  );
+
+  const handleEditorDidMount = (editor) => (editorRef.current = editor);
+  const handleEditorChange = (value: string | undefined) => {
+    value && setData(JSON.parse(value));
+  };
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/JSONViewer.tsx:36]: nit: I noticed that across couple pages there are same code for gettingthese values/functions. You can abstract these into a single hook useCustomColors() that return all these values and functions. then you can use it like

const { colorBy, useMulticolorKeypoints, etc } = useCustomColors()

File: app/packages/core/src/components/ColorModal/JSONViewer.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    });
+  }, []);
+
+  if (!global) return null;
+  return (
+    <div style={{ width: "100%", height: "100%", overflow: "hidden" }}>
+      <SectionWrapper>
+        You can use the JSON editor to customize the color settings.
+        <ActionOption
+          href={SORT_BY_SIMILARITY}
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/JSONViewer.tsx:64]: nit: does this link actually have content related to this yet?

File: app/packages/core/src/components/ColorModal/ShareStyledDiv.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  flex: 1,
+  textAlign: "center",
+};
+
+export const LONG_BUTTON_STYLE: React.CSSProperties = {
+  margin: "0.5rem 1rem",
+  height: "2rem",
+  flex: 2,
+  textAlign: "center",
+};
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/ShareStyledDiv.ts:112]: nit: these two are the same. You can pass flex as a parameter to one and refactor

File: app/packages/core/src/components/ColorModal/ShareStyledDiv.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  cursor: pointer;
+  font-weight: 600;
+`;
+
+export const Text = styled.div`
+  font-size: 1.1rem;
+  text-transform: uppercase;
+  width: 100%;
+  display: flex;
+  flex-direction: row;
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/ShareStyledDiv.ts:58]: nit: if default is row, we can eliminate all `flex-direction: row;`

File: app/packages/core/src/components/ColorModal/colorPalette/AttributeColorSetting.tsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    color: coloring.pool[Math.floor(Math.random() * coloring.pool.length)],
+  };
+  const [tempColor, setTempColor] = useRecoilState(tempColorSetting);
+  const values = tempColor.labelColors;
+
+  const [showPicker, setShowPicker] = useState(
+    Array(values?.length ?? 0).fill(false)
+  );
+
+  const handleAdd = () => {
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/colorPalette/AttributeColorSetting.tsx:65]: nit: useMemo?

File: app/packages/core/src/components/ColorModal/GlobalSetting.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  const opacity = useRecoilValue(fos.alpha(false));
+  const colorBy = useRecoilValue(
+    fos.appConfigOption({ key: "colorBy", modal: false })
+  );
+  const useMulticolorKeypoints = useRecoilValue(
+    fos.appConfigOption({ key: "multicolorKeypoints", modal: false })
+  );
+  const showSkeleton = useRecoilValue(
+    fos.appConfigOption({ key: "showSkeletons", modal: false })
+  );
```
Comment (manivoxel51) [inline app/packages/core/src/components/ColorModal/GlobalSetting.tsx:33]: nit: same here, similar code across at least 3/4 files, would be nice to refactor these into a hook

File: app/packages/core/src/components/FieldLabelAndInfo/index.tsx [hunk 11]
```diff
@@ -273,6 +293,13 @@ function FieldInfoExpanded({
       onClick={(e) => e.stopPropagation()}
     >
       <FieldInfoExpandedContainer color={color}>
+        {field.embeddedDocType && !isModal && canEdit && (
```
Comment (manivoxel51) [inline app/packages/core/src/components/FieldLabelAndInfo/index.tsx:296]: this is basically one of the few places we are adding our feature 👍

File: app/packages/looker/src/elements/common/tags.ts [hunk 12]
```diff
@@ -426,3 +423,58 @@ const getFieldAndValue = (
 
   return [field, value, list];
 };
+
+const compareObjectArrays = (arr1, arr2) => {
```
Comment (manivoxel51) [inline app/packages/looker/src/elements/common/tags.ts:427]: nit: can you use loadash for these?

File: app/packages/looker/src/elements/common/tags.ts [hunk 13]
```diff
@@ -426,3 +423,58 @@ const getFieldAndValue = (
...
+      }
+    }
+  }
+
+  // If all objects pass the comparison checks, return true
+  return true;
+};
+
+// Helper function to sort arrays of objects based on their key-value pairs
+function sortObjectArrays(a, b) {
```
Comment (manivoxel51) [inline app/packages/looker/src/elements/common/tags.ts:463]: nit: same here

File: app/packages/looker/src/overlays/base.ts [hunk 14]
```diff
@@ -93,20 +93,50 @@ export abstract class CoordinateOverlay<
...
+  }: Readonly<State>): string {
     let key;
+    let pool;
 
     switch (coloring.by) {
       case "field":
+        // check if the field has a customized color, use it if it is a valid color
+        const fieldColor = customizeColorSetting.find(
+          (s) => s.field === this.field
+        )?.fieldColor;
```
Comment (manivoxel51) [inline app/packages/looker/src/overlays/base.ts:107]: nit: I am unfamiliar with this code. hopefully the computation here is small enough not to impact rendering performance

File: fiftyone/server/mutation.py [hunk 15]
```diff
@@ -391,6 +408,35 @@ async def set_spaces(
...
+
+        if save_to_app:
+            view._dataset.app_config.color_scheme = foo.ColorSchemeDocument(
+                color_pool=color_scheme_save_format.color_pool,
+                customized_color_settings=color_scheme_save_format.customized_color_settings,
+            )
+            view._dataset.save()
+            state.view = view
+
+        await dispatch_event(subscription, StateUpdate(state=state))
```
Comment (manivoxel51) [inline fiftyone/server/mutation.py:437]: you might have to dispatch if save_to_app if you are following save view approach

**Review by manivoxel51 — APPROVED** [review]
nice work @lanzhenw ! 🍨 

The code is looking good! nice work separating the logic between different options in color settings 💯 

it looks reasonably safe and isolated. Let's merge this if
- basic app features are working as expected - for example, you can open sample modal, tag, navigate between them, filtering, visualization, etc. 

for followup: One trend I noticed was filtering or mapping over collections and not caching them. This can cause performance issues if heavy calculations run for every field every re-render. We can avoid some these maybe by adding useMemos, useCallbacks, etc. to avoid re-defining these values/functions on re-renders

File: app/packages/core/src/components/ColorModal/ColorModal.tsx [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    fos.fields({
+      space: fos.State.SPACE.SAMPLE,
+      ftype: EMBEDDED_DOCUMENT_FIELD,
+    })
+  )?.filter((f) =>
+    VALID_LABEL_TYPES.includes(
+      f?.embeddedDocType?.split(".")?.slice(-1)[0] ?? ""
+    )
+  );
+  const customizeColorFields = [...videoFields, ...sampleFields];
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/core/src/components/ColorModal/ColorModal.tsx:68]: I made a `eligibleFieldsToCustomizeColor` recoil selector.

File: app/packages/state/src/hooks/useClearSessionColorScheme.ts [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+import {
+  stateSubscription,
+  sessionColorScheme,
+  datasetName,
+  view,
+} from "../recoil";
+import useSendEvent from "./useSendEvent";
+import { DEFAULT_APP_COLOR_SCHEME } from "../utils";
+
+const useClearSessionColorScheme = () => {
```
[the remark this answers is not in the document]
Comment (lanzhenw) [reply app/packages/state/src/hooks/useClearSessionColorScheme.ts:14]: Implemented.

File: app/packages/looker/src/elements/common/tags.ts (hunk 12 above)
Comment (lanzhenw) [reply app/packages/looker/src/elements/common/tags.ts:427]: the lodash `isEqual` function is strict with the order of the array. These two helper functions only check if elements are the same, regardless of order.

**Pull request merged**